@xen-orchestra/backups/backupWorker.js

```diff
@@ -0,0 +1,162 @@
+
+  const worker = fork(path.resolve(__dirname, 'backupWorker.js'))
+
+  worker.on('exit', code => reject(new Error(`worker exited with code ${code}`)))
+  worker.on('error', reject)
+
+  worker.on('message', log => {
+    try {
+      if (log.workerEnd) {
+        if (log.error !== undefined) {
+          reject(log.error)
+        } else {
+          resolve(log.result)
+        }
+      } else {
+        onLog(log)
+      }
+    } catch (error) {
+      warn(error)
+    }
+  })
+
+  worker.send({
+    runWithLogs: onLog !== undefined,
+    ...params,
+  })
+
+  return promise
+}
+
+class BackupWorker {
+  #config
+  #job
+  #recordToXapi
+  #remotes
+  #schedule
+  #xapis
+
+  constructor({ config, job, recordToXapi, remotes, schedule, xapis }) {
+    this.#config = config
+    this.#job = job
+    this.#recordToXapi = recordToXapi
+    this.#remotes = remotes
+    this.#schedule = schedule
+    this.#xapis = xapis
+
+    const debounceResource = createDebounceResource()
+    debounceResource.defaultDelay = parseDuration(config.resourceCacheDelay)
+    this.debounceResource = debounceResource
+  }
+
+  run() {
+    return new Backup({
+      config: this.#config.backups,
+      getAdapter: remoteId => this.getAdapter(this.#remotes[remoteId]),
+      getConnectedRecord: Disposable.factory(async function* getConnectedRecord(type, uuid) {
+        const xapiId = this.#recordToXapi[uuid]
+        if (xapiId === undefined) {
+          throw new Error('no XAPI associated to ' + uuid)
+        }
+
+        const xapi = yield this.getXapi(this.#xapis[xapiId])
+        return xapi.getRecordByUuid(type, uuid)
+      }).bind(this),
+      job: this.#job,
+      schedule: this.#schedule,
+    }).run()
+  }
+
+  getAdapter = Disposable.factory(this.getAdapter)
+  getAdapter = deduped(this.getAdapter, remote => [remote.url])
+  getAdapter = compose(this.getAdapter, function (resource) {
+    return this.debounceResource(resource)
+  })
+  async *getAdapter(remote) {
+    const config = this.#config
+    const handler = getHandler(remote, config.remoteOptions)
+    await handler.sync()
+    try {
+      yield new RemoteAdapter(handler, {
+        debounceResource: this.debounceResource,
+        dirMode: config.backups.dirMode,
+      })
+    } finally {
+      await handler.forget()
+    }
+  }
+
+  getXapi = Disposable.factory(this.getXapi)
+  getXapi = deduped(this.getXapi, ({ url }) => [url])
+  getXapi = compose(this.getXapi, function (resource) {
+    return this.debounceResource(resource)
+  })
+  async *getXapi({ credentials: { username: user, password }, ...opts }) {
+    const xapi = new Xapi({
+      ...this.#config.xapiOptions,
+      ...opts,
+      auth: {
+        user,
+        password,
+      },
+    })
+
+    await xapi.connect()
+    try {
+      await xapi.objectsFetched
+
+      yield xapi
+    } finally {
+      await xapi.disconnect()
+    }
+  }
+}
+
+process.on('message', async ({ runWithLogs, ...rest }) => {
```

Comment:
What if the process receive multliple messages?

Define a proper protocol between the runner and the worker.

Reply:
It cannot receive multiple message, an instance of this file is created per execution.

Reply:
We will need to send messages to running job in a soon future (eg cancelation), we have to implement a basic message system here.

Reply:
I don't know how the cancellation will be implemented.

 Do you have an example of cancellation code that will help me to suggest an appropriate solution for your need?

Reply:
No matter how it will be implemented, the cancellation will be triggered in the running code, and it will need to be propagated to the backup code.

For that, we'll probably use a simple message for this propagation, something like:

```js
{ action: "cancel", reason: "cancellation message" }
```